File: audit/audit.py
import json
import subprocess
import re


def empty(value):
    try:
        value = float(value)
    except ValueError:
        pass
    return bool(value)


def read_file(filename):
    try:
        return open(filename).read().strip()
    except:
        "not available"


def read_command(command):
    try:
        res = subprocess.check_output(command.split()).strip()
        return res
    except OSError:
        return "not available"
    except:
        return "not available"


def get_sysctl():
    i = {}
    for item in filter(None, read_command('sysctl -a').split("\n")):
        if " = " in item:
            j = item.split(" = ")
        elif ": " in item:
            j = item.split(": ")

        i[j[0].strip()] = j[1].strip()

    return i


def get_release():
    rels = ["/etc/SuSE-release", "/etc/redhat-release", "/etc/redhat_version", "/etc/fedora-release",
        "/etc/slackware-release", "/etc/slackware-version", "/etc/debian_release", "/etc/debian_version",
        "/etc/os-release", "/etc/mandrake-release", "/etc/yellowdog-release", "/etc/sun-release",
        "/etc/release", "/etc/gentoo-release", "/etc/system-release", "/etc/lsb-release"]
    for rel in rels:
        try:
            res = read_file(rel)
        except:
            pass

    return res


def get_scheduler():
    try:
        p1 = subprocess.Popen(['ls', '-l', '/sys/block'], stdout=subprocess.PIPE)
        p2 = subprocess.Popen(['awk', "{print $9}"], stdin=p1.stdout, stdout=subprocess.PIPE)
        blocks = filter(None, p2.communicate()[0].split("\n"))

        res = {}
        for block in blocks:
          res[block] = read_file('/sys/block/' + block + '/queue/scheduler')

        return res
    except:
        return "not available"


def tabular_data(data):
    res = []
    lines = data.split("\n")

    # keys
    first_line = re.sub('Local Address:Port','Local-Address-Port', lines[0])
    first_line = re.sub('Peer Address:Port','Peer-Address-Port', first_line)
    keys = ' '.join(first_line.split()).split(' ')

    try:
        keys.remove('on')
    except:
        pass

    for i, s in enumerate(keys):
        keys[i] = re.sub(r'\W+', '', s.lower().strip())

    # data
    for i in range(1,len(lines)):
        a = {}
        vals = ' '.join(lines[i].split()).split(' ')

        # /proc/partitions has no data in the second row.  If no val present, skip the row
        if vals[0] == "":
          continue

        for j in range(len(keys)):
            key = keys[j]
            try:
                val = vals[j]
            except:
                val = ''
            a[key] = val
        res.append(a)

    return res


def delimited_data(delimiter, data):
    res = {}
    lines = filter(None, data.split("\n"))
    for i in range(len(lines)):
        j = lines[i].split(delimiter)
        res[re.sub(r'\W+', '_', j[0].lower().strip())] = j[1].strip()

    return res


def main():

    a = {}

    # SYSCTL
    a['sysctl'] = get_sysctl()

    # PROC
    a['proc'] = {}
    a['proc']['cpuinfo'] = delimited_data(':', read_file('/proc/cpuinfo'))
    a['proc']['cmdline'] = read_file('/proc/cmdline')
    a['proc']['net/softnet_stat'] = read_file('/proc/net/softnet_stat')
    a['proc']['cgroups'] = tabular_data(read_file('/proc/cgroups'))
    a['proc']['uptime'] = read_file('/proc/uptime')
    a['proc']['vmstat'] = delimited_data(' ', read_file('/proc/vmstat'))
    a['proc']['loadavg'] = read_file('/proc/loadavg')
    a['proc']['zoneinfo'] = read_file('/proc/zoneinfo')
    a['proc']['partitions'] = tabular_data(read_file('/proc/partitions'))
    a['proc']['version'] = read_file('/proc/version')

    # PARTITIONS
    a['disk_partitions'] = tabular_data(read_command('df -h'))

    # DMESG
    a['dmesg'] = read_command('dmesg')

    # THP
    a['transparent_huge_pages'] = {}
    a['transparent_huge_pages']['enabled'] = read_file('/sys/kernel/mm/transparent_hugepage/enabled')
    a['transparent_huge_pages']['defrag'] = read_file('/sys/kernel/mm/transparent_hugepage/defrag')

    # MEMORY
    a['memory'] = read_command('free -m')

    # DISK
    a['disk'] = {}
    a['disk']['scheduler'] = get_scheduler()
    a['disk']['number_of_disks'] = tabular_data(read_command('lsblk'))

    # NETWORK
    a['network'] = {}
    a['network']['ifconfig'] = read_command('ifconfig')
    a['network']['ip'] = read_command('ip addr show')
    a['network']['netstat'] = read_command('netstat -a')
    a['network']['ss'] = tabular_data(read_command('ss -tan'))

    # DISTRO
    a['distro'] = {}
    a['distro']['issue'] = read_file('/etc/issue')
    a['distro']['release'] = get_release()

    # POWER MGMT
    a['power_mgmt'] = {}
    a['power_mgmt']['max_cstate'] = read_file('/sys/module/intel_idle/parameters/max_cstate')

    print json.dumps(a, indent=4)


if __name__ == "__main__":
   main()

File: audit/audit.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// NotAvailable human readable
var NotAvailable = "not available"

// Audit holds the key settings of the system
type Audit struct {
	Sysctl    map[string]string `json:"sysctl"`
	Proc      Proc              `json:"proc"`
	Dmesg     string            `json:"dmesg"`
	THP       THP               `json:"transparent_huge_pages"`
	Memory    string            `json:"memory"`
	Disk      Disk              `json:"disk"`
	Network   Network           `json:"network"`
	Distro    Distro            `json:"distro"`
	PowerMgmt PowerMgmt         `json:"power_mgmt"`
}

// Proc holds the key settings of proc interface
type Proc struct {
	Cpuinfo        string            `json:"cpuinfo"`
	Cmdline        string            `json:"cmdline"`
	NetSoftnetStat string            `json:"net/softnet_stat"`
	Cgroups        string            `json:"cgroups"`
	Uptime         string            `json:"uptime"`
	Vmstat         map[string]string `json:"vmstat"`
	Loadavg        string            `json:"loadavg"`
	Zoneinfo       string            `json:"zoneinfo"`
	Partitions     string            `json:"partitions"`
	Version        string            `json:"version"`
}

// THP handles the transparent huge pages
type THP struct {
	Enabled string `json:"enabled"`
	Defrag  string `json:"defrag"`
}

// Disk handles the disk subystem
type Disk struct {
	Scheduler  map[string]string `json:"scheduler"`
	NumDisks   string            `json:"number_of_disks"`
	Partitions string            `json:"partitions"`
}

// Network handles the networking settings
type Network struct {
	Ifconfig string `json:"ifconfig"`
	IP       string `json:"ip"`
	Netstat  string `json:"netstat"`
	SS       string `json:"ss"`
}

// Distro handles the linux distro settings
type Distro struct {
	Issue   string `json:"issue"`
	Release string `json:"release"`
}

// PowerMgmt handles the power management settings
type PowerMgmt struct {
	MaxCState string `json:"max_cstate"`
}

// getSysctl return output from sysctl command
func getSysctl() map[string]string {
	out, err := exec.Command("sysctl", "-a").Output()
	kv := make(map[string]string)
	if err == nil {
		lines := strings.Split(string(out), "\n")
		for _, pair := range lines {

			if strings.ContainsAny(pair, ":") {
				z := strings.Split(pair, ":")
				kv[z[0]] = strings.TrimSpace(z[1])
			}

			if strings.ContainsAny(pair, "=") {
				z := strings.Split(pair, "=")
				kv[z[0]] = strings.TrimSpace(z[1])
			}
		}
	}
	return kv
}

// read_command captures the output of a command
func read_command(name string, arg ...string) string {
	out, err := exec.Command(name, arg...).Output()
	if err != nil {
		return NotAvailable
	}
	return strings.TrimSpace(string(out))
}

// read_file captures the contents of a file
func read_file(filename string) string {
	dat, err := ioutil.ReadFile(filename)
	if err != nil {
		return NotAvailable
	}
	return strings.TrimSpace(string(dat))
}

// delimited_data splits data by a delimiter
func delimited_data(delimiter string, data string) map[string]string {
	kv := make(map[string]string)

	if strings.Contains(data, NotAvailable) {
		return kv
	}

	lines := strings.Split(data, "\n")
	for _, pair := range lines {
		if strings.ContainsAny(pair, delimiter) {
			z := strings.Split(pair, delimiter)
			key := strings.TrimSpace(z[0])
			value := strings.TrimSpace(z[1])
			kv[key] = value
		}
	}
	return kv
}

// get_release checks with linux distro it is
func get_release() string {
	rels := []string{
		"/etc/SuSE-release", "/etc/redhat-release", "/etc/redhat_version",
		"/etc/fedora-release", "/etc/slackware-release",
		"/etc/slackware-version", "/etc/debian_release", "/etc/debian_version",
		"/etc/os-release", "/etc/mandrake-release", "/etc/yellowdog-release",
		"/etc/sun-release", "/etc/release", "/etc/gentoo-release",
		"/etc/system-release", "/etc/lsb-release",
	}
	for _, path := range rels {
		if _, err := os.Stat(path); err == nil {
			return read_file(path)
		}
	}
	return NotAvailable
}

// get_scheduler handles capturing scheduler data for each block device
func get_scheduler() map[string]string {
	kv := make(map[string]string)
	files, err := ioutil.ReadDir("/sys/block")
	if err != nil {
		return kv
	}
	for _, f := range files {
		block := f.Name()
		path := "/sys/block/" + block + "/queue/scheduler"
		if _, err := os.Stat(path); err == nil {
			kv[block] = read_file(path)
		}
	}
	return kv
}

func main() {
	uid := os.Getuid()

	if uid != 0 {
		log.Fatal("This script must be run as the user root")
	}

	jsonData := Audit{
		Sysctl: getSysctl(),
		Proc: Proc{
			Cpuinfo:        read_file("/proc/cpuinfo"),
			Cmdline:        read_file("/proc/cmdline"),
			NetSoftnetStat: read_file("/proc/net/softnet_stat"),
			Cgroups:        read_file("/proc/cgroups"),
			Uptime:         read_file("/proc/uptime"),
			Vmstat:         delimited_data(" ", read_file("/proc/vmstat")),
			Loadavg:        read_file("/proc/loadavg"),
			Zoneinfo:       read_file("/proc/zoneinfo"),
			Partitions:     read_file("/proc/partitions"),
			Version:        read_file("/proc/version"),
		},
		Dmesg: read_command("dmesg"),
		THP: THP{
			Enabled: read_file("/sys/kernel/mm/transparent_hugepage/enabled"),
			Defrag:  read_file("/sys/kernel/mm/transparent_hugepage/defrag"),
		},
		Memory: read_command("free", "-m"),
		Disk: Disk{
			Scheduler:  get_scheduler(),
			Partitions: read_command("df", "-h"),
			NumDisks:   read_command("lsblk"),
		},
		Network: Network{
			Ifconfig: read_command("ifconfig"),
			IP:       read_command("ip", "addr", "show"),
			Netstat:  read_command("netstat", "-an"),
			SS:       read_command("ss", "-tan"),
		},
		Distro: Distro{
			Issue:   read_file("/etc/issue"),
			Release: get_release(),
		},
		PowerMgmt: PowerMgmt{
			MaxCState: read_file("/sys/module/intel_idle/parameters/max_cstate"),
		},
	}
	b, _ := json.MarshalIndent(jsonData, "", "  ")
	hostname, _ := os.Hostname()
	timestamp := strconv.FormatInt(time.Now().Unix(), 10)
	auditFile := fmt.Sprintf("audit-%s-%s.json", hostname, timestamp)

	err := ioutil.WriteFile(auditFile, b, 0644)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Generated audit file: %s\n", auditFile)
}
